File: MapReduce_FastQ_BaseQual_Optimized/src/main/java/baseQual/FastQ_PerBaseQual_Mapper.java
package baseQual;

import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.lib.input.FileSplit;
import org.seqdoop.hadoop_bam.SequencedFragment;
import utils.IdentifierPositionKeyWritable;
import utils.QualityCountHelperWritable;

import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 04.02.16.
 */
public class FastQ_PerBaseQual_Mapper extends org.apache.hadoop.mapreduce.Mapper<Object, SequencedFragment, IdentifierPositionKeyWritable, QualityCountHelperWritable> {
    private static final int OFFSET = 33;

    private Map<IdentifierPositionKeyWritable, QualityCountHelperWritable> keyMapping = new HashMap<IdentifierPositionKeyWritable, QualityCountHelperWritable>();

    @Override
    protected void map(Object key, SequencedFragment value, Context context) throws IOException, InterruptedException {
        final String sampleIdentifier = ((FileSplit)context.getInputSplit()).getPath().getName();

        final Text quality = value.getQuality();
        for (int i = 0; i < quality.getLength(); i++) {
            IdentifierPositionKeyWritable outputKey = new IdentifierPositionKeyWritable(sampleIdentifier, (i + 1));
            int qualValue = getCorrespondingIntValue(quality.charAt(i));

            if(!this.keyMapping.containsKey(outputKey)) {
                this.keyMapping.put(outputKey, new QualityCountHelperWritable(qualValue));
            } else {
                this.keyMapping.get(outputKey).addQualityValueToSum(qualValue);
                this.keyMapping.get(outputKey).incrementNumberOfQualityValues();
            }
        }
    }

    @Override
    protected void cleanup(Context context) throws IOException, InterruptedException {
        Iterator it = this.keyMapping.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry pair = (Map.Entry)it.next();
            context.write((IdentifierPositionKeyWritable) pair.getKey(), (QualityCountHelperWritable) pair.getValue());
        }
    }

    private int getCorrespondingIntValue(int qualVal) {
        return qualVal - FastQ_PerBaseQual_Mapper.OFFSET;
    }

}

File: Spark_FastQ_BaseQual_SingleFile_Baseline/src/main/scala/FastQ_PerBaseQual_Mapper.scala
package main.scala

import main.scala.utils.AvgCount
import org.apache.hadoop.io.Text
import org.seqdoop.hadoop_bam.SequencedFragment

import scala.collection.mutable.ListBuffer

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 21.04.2016
  */
object FastQ_PerBaseQual_Mapper {
  private val OFFSET: Int = 33

  def flatMap(record: SequencedFragment): TraversableOnce[Pair[Int, Int]] = {
    val resList = new ListBuffer[Pair[Int, Int]]()
    val quality: Text = record.getQuality
    for(i <- 0 until quality.getLength) {
      resList.append(new Pair((i+1), getCorrespoindingIntValue(quality.charAt(i))))
    }
    resList.toTraversable
  }

  def map(pair: Pair[Int, AvgCount]): Pair[Int, Double] = {
    new Pair(pair._1, pair._2.getMeanValue())
  }

  private def getCorrespoindingIntValue(qualVal: Int): Int = {
    (qualVal - OFFSET)
  }

}